### ISSUE TITLE
Training specific blocks?

### ISSUE THREAD

**Issue by cchance27** [tdrussell/diffusion-pipe#66]
Is it possible to focus training on specific blocks, i know this is something thats been done in some others, but after seeing on civitai someone mentioned to combine loras, they were horrible with 2 or more loras unless they disabled single blocks, and double 0-9... 

I recall with other models disabling excess blocks, or training a specific layer was MUCH faster, so if it seems that singles are less useful for person loras, wouldn't this drastically speed up training if we could just train the target double blocks?

**Comment by alisson-anjos** [comment]
If you want to take a look at the implementation, I made the changes to train only in double blocks in my fork https://github.com/alisson-anjos/diffusion-pipe-ui/blob/68b0a02d1d0cf9177cb8bf2f899fcd5fd11f88f2/models/base.py#L124 I added a only_double_blocks configuration in the configuration file in the adapter section.

**Comment by rkfg** [comment]
Very interesting, I just thought about this optimization too since full lora merge causes a blocky output (ironically) if there's more than one lora. Only applying double blocks fixes it. However, if you only train the double blocks wouldn't it attribute the entire loss to them instead of spreading it between single and double blocks? I suppose the result would be different, maybe you'll need to lower LR as well or do other tricks. Like, calculate the single block loss too but don't propagate it further and don't apply it.

**Comment by rkfg** [comment]
I applied the change to `diffusion-pipe` (this repo):
```diff
diff --git a/models/base.py b/models/base.py
index 3d910fc..9c0557f 100644
--- a/models/base.py
+++ b/models/base.py
@@ -120,9 +120,12 @@ class BasePipeline:
 
     def configure_adapter(self, adapter_config):
         target_linear_modules = set()
+        only_double_blocks = adapter_config.get('only_double_blocks', False)
         for name, module in self.transformer.named_modules():
             if module.__class__.__name__ not in self.adapter_target_modules:
                 continue
+            if only_double_blocks and "single_blocks" in name:
+                continue
             for full_submodule_name, submodule in module.named_modules(prefix=name):
                 if isinstance(submodule, nn.Linear):
                     target_linear_modules.add(full_submodule_name)
```
Then I added `only_double_blocks = true` to the `[adapter]` section. The speedup is quite small to my surprise, maybe 5-10% tops. But the resulting lora is 173 MB (rank 32) instead of 300+ IIRC, so that's already a good improvement. The quality seems to be good as well, only trained 15 epochs for a character, might do more.
